src/api/objects/ObjectAPI.js

```diff
@@ -20,337 +20,387 @@
-     * and so forth.
-     *
-     * A few common properties are defined for domain objects. Beyond these,
-     * individual types of domain objects may add more as they see fit.
-     *
-     * @property {module:openmct.ObjectAPI~Identifier} identifier a key/namespace pair which
-     *           uniquely identifies this domain object
-     * @property {string} type the type of domain object
-     * @property {string} name the human-readable name for this domain object
-     * @property {string} [creator] the user name of the creator of this domain
-     *           object
-     * @property {number} [modified] the time, in milliseconds since the UNIX
-     *           epoch, at which this domain object was last modified
-     * @property {module:openmct.ObjectAPI~Identifier[]} [composition] if
-     *           present, this will be used by the default composition provider
-     *           to load domain objects
-     * @typedef DomainObject
-     * @memberof module:openmct
-     */
-
-    function hasAlreadyBeenPersisted(domainObject) {
-        return domainObject.persisted !== undefined
-            && domainObject.persisted === domainObject.modified;
+            domainObject.persisted = persistedTime;
+            provider.create(domainObject).then((response) => {
+                this.mutate(domainObject, 'persisted', persistedTime);
+                savedResolve(response);
+            });
+        } else {
+            domainObject.persisted = persistedTime;
+            this.mutate(domainObject, 'persisted', persistedTime);
+            result = provider.update(domainObject);
+        }
+    }
+
+    return result;
+};
+
+/**
+ * Add a root-level object.
+ * @param {module:openmct.ObjectAPI~Identifier|function} an array of
+ *        identifiers for root level objects, or a function that returns a
+ *        promise for an identifier or an array of root level objects.
+ * @method addRoot
+ * @memberof module:openmct.ObjectAPI#
+ */
+ObjectAPI.prototype.addRoot = function (key) {
+    this.rootRegistry.addRoot(key);
+};
+
+/**
+ * Modify a domain object.
+ * @param {module:openmct.DomainObject} object the object to mutate
+ * @param {string} path the property to modify
+ * @param {*} value the new value for this property
+ * @method mutate
+ * @memberof module:openmct.ObjectAPI#
+ */
+ObjectAPI.prototype.mutate = function (domainObject, path, value) {
+    if (!this.supportsMutation(domainObject)) {
+        throw `Error: Attempted to mutate immutable object ${domainObject.name}`;
+    }
+
+    if (domainObject.isMutable) {
+        domainObject.$set(path, value);
+    } else {
+        //Creating a temporary mutable domain object allows other mutable instances of the
+        //object to be kept in sync.
+        let mutableDomainObject = this._toMutable(domainObject);
+
+        //Mutate original object
+        MutableDomainObject.mutateObject(domainObject, path, value);
```

Comment:
Do we need to call`MutableDomainObject.mutateObject(domainObject, path, value);` before we call `mutableDomainObject.$set(path, value)`? 
It appears that mutableDomainObject.$set already does what the static method does here: https://github.com/nasa/openmct/pull/3483/files?file-filters%5B%5D=.js&hide-deleted-files=true#diff-7ef3a5358f05632acb4ff900098f346817a36e1759a4dc36972629b36afeddd2R72

Reply:
The problem is that creating a `MutableDomainObject` actually creates a copy of the non-mutable `domainObject` that is passed in. So calling `$set` on it will mutate the copy, and emit an event so that any other `MutableDomainObject` instances of the same object get updated, but it will not actually mutate the original `domainObject`. 

"Why does it create a copy?" you might very reasonably ask. The aim was to stick to the principle that domain object are just plain old JavaScript objects, without any functions or special properties added by the framework. This was one of the foundational principles of the refactor away from the original API.

It achieves this via inheritance, which requires a new object to be created that inherits the special functions and properties of the `MutableDomainObject` without them existing on the object itself.